specs/guaranteed-gas-market.md

```diff
@@ -0,0 +1,137 @@
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+[Deposited transaction](./glossary.md#deposited-transaction) are transactions on L2 that are
+initiated on L1. The gas that they use on L2 is bought on L1 via a gas burn or a direct payment. We
+maintain a fee market and hard cap on the amount of gas provided to all deposits in a single L1
+block.
+
+The gas provided to deposited transactions is sometimes called "guaranteed gas". The gas provided to
+deposited transactions is unique in the regard that it is not refundable. It cannot be refunded as
+it is sometimes paid for with a gas burn and there may not be any ETH left to refund.
+
+The **guaranteed gas** is composed of a gas stipend, and of any guaranteed gas the user would like
+to purchase (on L1) on top of that.
+
+Guaranteed gas on L2 is bought in the following manner. An L2 gas price is calculated via a 1559
+style gas market. The total amount of ETH required to buy that gas is then calculated
+(`guaranteed gas * L2 deposit basefee`). The contract then accepts that amount of ETH (in a future
+upgrade) or (only method right now), burns an amount of L1 gas that corresponds to the L2 cost
+(`L2 cost / L1 Basefee`).
+
+## Gas Stipend
+
+To offset the gas spent on the deposit event, we credit the amount of gas spent on the metering step
+times the current basefee to the cost of the L2 gas. The amount of gas is selected to represent the
+cost to the user. If the ETH price of the gas (gas times current L1 baseefee) is greater than the
+requested guaranteed gas times the L2 gas price, no L1 gas is burnt.
+
+## Limiting Guaranteed Gas
+
+The total amount of guaranteed gas that can be bought in a single L1 block must be limited to
+prevent a denial of service attack against L2 as well as allow the total amount of guaranteed gas
+to be below the L2 block gas limit.
+
+We set a guaranteed gas limit of 8,000,000 gas per L1 block and a target of 2,000,000 gas per L1
+block. These numbers enabled occasional large transactions while staying within our target and
+maximum gas usage on l2.
+
+## 1559 Fee Market
+
+To reduce [Priority Gas Auctions](./glossary.md#priority-gas-auction) and accurately price gas, we
+implement a 1559 style fee market on L1 with the following pseudocode. We also use this opporunity
+to place a hard limit on the amount of guaranteed gas that is provided.
+
+```python
+# Pseudocode to update the L2 Deposit Basefee and cap the amount of guaranteed gas
+# bought in a block. Calling code must handle the gas burn and validity checks on
+# the ability of the account to afford this gas.
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+ELASTICITY_MULTIPLIER = 4
+MAX_RESOURCE_LIMIT = 8,000,000
+TARGET_RESOURCE_LIMIT = MAX_RESOURCE_LIMIT / ELASTICITY_MULTIPLIER
+MINIMUM_BASEFEE=10000
+
+# prev_basefee is a u128, prev_bought_gas and prev_num are u64s
+prev_basefee, prev_bought_gas, prev_num = load_and_unpack_storage()
```

Comment:
Isn't this changing to be their own slots?

Reply:
We place them into a struct which solidity packs into a single slot.

Reply:
https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/contracts/L1/ResourceMetering.sol

Reply:
Given that the solidity doesn't have `load_and_unpack_storage()`, just refers to the struct fields, could we just do something like:
`prev_basefee, prev_bought_gas, prev_num = <values from previous update>` ?

Or just have top-level `prev_basefee, prev_bought_gas, prev_num` vars with proper initial value, then change the algorithm to be contained with an `update` function that takes some parameters (`now_num`, `requested_gas` ... maybe I'm missing something else). At the end of the function assign `prev_basefee, prev_bought_gas, prev_num` to their proper values.

My objections here is that `load_and_unpack_storage()` is more like an implementation detail that doesn't have a *semantic* meaning in the algorithm. It doesn't really matter how we store the values to compute the amount of gas, this is a platonic algorithm which can then be impemented on chain or even somewhere else for simulation purpose.